receiver/snmpreceiver/scraper.go

```diff
@@ -0,0 +1,559 @@
+	// metricsByResourceMap is provided as an easy way to check if a metric is already
+	// associated with an existing resource
+	scalarMetricCnt := len(metricsMap)
+	if err := s.scrapeIndexedMetrics(now, resourceMetricsMap, metricsByResourceMap); err != nil {
+		s.logger.Warn("Failed to collect column OID metrics", zap.Error(err))
+	} else if len(metricsByResourceMap[""]) > scalarMetricCnt || len(metricsByResourceMap) > 1 {
+		metricsProcessed = true
+	}
+
+	// Return error if we failed to scrape any metrics
+	md := pmetric.NewMetrics()
+	if !metricsProcessed {
+		return md, errScrape
+	}
+
+	// Put all of the created ResourceMetrics onto a top level Metrics
+	for _, value := range resourceMetricsMap {
+		value.MoveTo(md.ResourceMetrics().AppendEmpty())
+	}
+
+	return md, nil
+}
+
+// scrapeScalarMetrics retrieves all SNMP data from scalar OIDs and turns the returned scalar data
+// into metrics with optional enum attributes
+func (s *snmpScraper) scrapeScalarMetrics(now pcommon.Timestamp, metricsMap map[string]*pmetric.Metric) (err error) {
+	scalarMetricNamesByOID := map[string]string{}
+	scalarMetricOIDs := []string{}
+
+	// Find all metric scalar OIDs
+	// Also create a map of metric names with OID as key so the metric config will be easy to
+	// matchup later with returned data
+	for name, metricCfg := range s.cfg.Metrics {
+		if len(metricCfg.ScalarOIDs) > 0 {
+			for i, oid := range metricCfg.ScalarOIDs {
+				// Data is returned by the client with '.' prefix on the OIDs.
+				// Making sure the prefix exists here in the configs so we can match it up with returned data later
+				if !strings.HasPrefix(oid.OID, ".") {
+					oid.OID = "." + oid.OID
+					s.cfg.Metrics[name].ScalarOIDs[i].OID = oid.OID
+				}
+				scalarMetricOIDs = append(scalarMetricOIDs, oid.OID)
+				scalarMetricNamesByOID[oid.OID] = name
+			}
+		}
+	}
+
+	// If no scalar metric configs, nothing else to do
+	if len(scalarMetricOIDs) == 0 {
+		return nil
+	}
+
+	// Get all SNMP scalar OID data and turn it into metrics/attributes
+	// which are then stored in the passed in metricsMap
+	err = s.client.GetScalarData(scalarMetricOIDs, scalarDataToMetric(now, metricsMap, scalarMetricNamesByOID, s.cfg))
+
+	return err
+}
+
+// scalarDataToMetric provides a function which will convert one piece of SNMP scalar data, turn it into
+// a metric with attributes based on the related configs, store it in the passed in metricsMap
+func scalarDataToMetric(
+	now pcommon.Timestamp,
+	metricsMap map[string]*pmetric.Metric,
+	scalarMetricNamesByOID map[string]string,
+	cfg *Config,
+) processFunc {
+	// This returns a processFunc because this is what the client's GetScalarData method requires
+	return func(data snmpData) error {
+		// Return an error if this SNMP scalar data is not of a useable type
+		switch data.valueType {
+		case NotSupported:
+			fallthrough
+		case String:
+			return fmt.Errorf(errMsgBadValueType, data.oid)
+		}
```

Comment:
nit: Wondering if this needs a switch seems like could just be an if statement.

Reply:
good callout. changed